docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+- Attester application
+    1. Generates an evidence (oe_get_evidence) of the next format from registered attesters and sends it to the verifier application.
+    2. Repeat until all formats are attempted or verification succeeds.
+
+Specification
+-------------
+
+## Differences between V2 and V3 APIs.
+
+Most of the APIs to attestation plugins are the same between the versions. The differences are highlighted below.
+
+## Changes
+
+### New APIs to query and select attesters and verifiers
+
+```C
+/**
+ * oe_sgx_get_attesters
+ * 
+ * Helper function that returns the SGX attesters that can then be sent to
+ * `oe_register_attester`.
+ *
+ * @param[out] attesters The SGX attesters that are available to the application.
+ * @param[out] attesters_length The length of the SGX attesters.
+ * @retval OE_OK on success.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_sgx_get_attesters(oe_attester_t** attesters, size_t *attesters_length);
+
+/**
+ * oe_free_attesters
+ *
+ * Frees the attesters.
+ *
+ * @param[in] attesters The attesters to be freed.
+ * @retval OE_OK on success.
+ */
+oe_result_t oe_free_attesters(oe_attester_t* attesters);
+
+/**
+ * oe_sgx_get_verifiers
+ * 
+ * Helper function that returns the SGX verifiers that can then be sent to
+ * `oe_register_verifier`.
+ *
+ * @param[out] verifiers The SGX verifiers that are available to the application.
+ * @param[out] verifiers_length The length of the SGX verifiers.
+ * @retval OE_OK on success.
+ */
+oe_result_t oe_sgx_get_verifiers(oe_verifier_t** verifiers, size_t *verifiers_length);
+
+/**
+ * oe_free_verifiers
+ *
+ * Frees the verifiers.
+ *
+ * @param[in] verifiers The list of the verifiers.
+ * @retval OE_OK on success.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_free_verifiers(oe_verifier_t* verifiers);
+
+/**
+ * oe_select_attester_format
+ *
+ * From the list, select an evidence format that the registered attesters
+ * can generate. If there's more than one match, the first one in the list
+ * wins.
+ *
+ * @param[in] format_ids The list of the format ids.
+ * @param[in] format_ids_length The length of the format ids list.
+ * @param[out] selected_format_id The selected id from the evidence
+ * format ids list.
+ * @retval OE_OK on success.
+ * @retval OE_NOT_FOUND if none of the format ids match any registered attesters.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_select_attester_format(
+    const oe_uuid_t* format_ids,
+    size_t format_ids_length,
+    oe_uuid_t** selected_format_id);
+
+/**
+ * oe_get_evidence_parameters
+ *
+ * Get evidence parameters from the verifier of the specified format id.
+ * These parameters are to be used by attester as optional inputs when generating evidence.
```

Comment:
I still have no idea what this API is for.

Reply:
Some evidence generation needs info from verifier as as input. For example, SGX local report. This step allows the verifier plugin to output parameters needed to generate evidence. For other evidence formats, this can be empty.

See the function definition below, the 5th parameter is an optional parameter that comes from the verifier:

oe_result_t oe_get_evidence(
    const oe_uuid_t* evidence_format_uuid,
    uint32_t flags,
    const oe_claim_t* custom_claims,
    size_t custom_claims_length,
    const void* opt_params,
    size_t opt_params_size,
    uint8_t** evidence_buffer,
    size_t* evidence_buffer_size,
    uint8_t** endorsements_buffer,
    size_t* endorsements_buffer_size);


---
In reply to: [393257161](https://github.com/openenclave/openenclave/pull/2621#discussion_r393257161) [](ancestors = 393257161)

Reply:
Suggest updating the text to use the nonce example discussed in IETF docs.

However, what I don't really understand yet is if a flat buffer is sufficient, instead of some structured set of parameters. I am afraid that in some attestation protocols each parameter has to be encoded separately in the protocol and a flat buffer might be problematic if so.   On the other hand, maybe the contract is that the buffer is expected to be in a particular format (as specified by the format_id) in which case it may be fine.  If the latter is correct, then it would be helpful to update the /* comments */ to explicitly say so.

Reply:
The format of the evidence parameters is a contract defined by the verifier and the attester of the format.

Comments updated.

---
In reply to: [394656908](https://github.com/openenclave/openenclave/pull/2621#discussion_r394656908) [](ancestors = 394656908)